### ISSUE TITLE
didReceiveIncomingPushWithPayload not called in background/terminated

### ISSUE THREAD

**Issue by aminosman** [react-native-webrtc/react-native-voip-push-notification#59]
Everything works fine when app is running in foreground, method gets called and call screen shows. If I minimize the app though nothing gets called and nothing gets logged even. I have read over the readme a million times and followed every step.

Read this:
https://developer.apple.com/library/ios/documentation/Performance/Conceptual/EnergyGuide-iOS/OptimizeVoIP.html

Added this:
`Background Modes -> Voice over IP`

Any thoughts? Running iOS 13.4.1 on iPhone XR w/ RN 0.61.5

**Comment by vsalle** [comment]
> Everything works fine when app is running in foreground, method gets called and call screen shows. If I minimize the app though nothing gets called and nothing gets logged even. I have read over the readme a million times and followed every step.
> 
> Read this:
> https://developer.apple.com/library/ios/documentation/Performance/Conceptual/EnergyGuide-iOS/OptimizeVoIP.html
> 
> Added this:
> `Background Modes -> Voice over IP`
> 
> Any thoughts? Running iOS 13.4.1 on iPhone XR w/ RN 0.61.5

Same behavior on my side

**Comment by aminosman** [comment]
Are you able to test on real device running any other version of iOS?

**Comment by VSPL** [comment]
I am also having same issue, read a lot but not finding solution. We have implemented this solution https://github.com/nimbleape/react-native-callkeep along with this solution. Our iOS scenario, 

In Foreground, 
Working fine

In Background, 
Not Working

In Killed State, 
Not Working

Please help.

**Comment by JJMoon** [comment]
I had same problem. It is because when the app is terminated, when the notification comes the JS part is not ready.
You can fire the event with 1 second delay as follows in the iOS code.

```
- (void)handleRemoteNotificationReceived:(NSNotification *)notification
dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
    [_bridge.eventDispatcher sendDeviceEventWithName:@"voipRemoteNotificationReceived"
                                                body:notification.userInfo];
});
}
```

**Comment by hoanganhnh2009** [comment]
> I had same problem. It is because when the app is terminated, when the notification comes the JS part is not ready.
> You can fire the event with 1 second delay as follows in the iOS code.
> 
> ```
> - (void)handleRemoteNotificationReceived:(NSNotification *)notification
> dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
>     [_bridge.eventDispatcher sendDeviceEventWithName:@"voipRemoteNotificationReceived"
>                                                 body:notification.userInfo];
> });
> }
> ```

Thanks for you.

**Comment by danj565** [comment]
I am having the exact same issue.  I tried @JJMoon 's suggestion but did not work (Thank you for the suggestion tho!).  I also tried the initialiationCompleted as linked by @JJMoon  which didn't work either.

2.1.0 / IPhone X / RN 0.62.2 / iOS 13.5.1 / XCode 11.5

@aminosman @vsalle - Did you find a solution?

Thank you!

**Comment by danj565** [comment]
I discovered the issue: https://stackoverflow.com/questions/58115125/ios-13-not-getting-voip-push-notifications-in-background

**Comment by adrian1388** [comment]
I am running into this issue in the part that JS is not running but not what @danj565 says, because I can see correct logs on didReceiveIncomingPushWithPayload on AppDelegate.

I tried @JJMoon answer but I have a problem with _bridge variable: "**Use of undeclared identifier '_bridge'**"

> ```
> - (void)handleRemoteNotificationReceived:(NSNotification *)notification
> dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
>     [_bridge.eventDispatcher sendDeviceEventWithName:@"voipRemoteNotificationReceived"
>                                                 body:notification.userInfo];
> });
> }
> ```

@JJMoon , could you please, explain me if this code is only available with your copy of the project? Sorry, My Objective C knowledge is not the best. How can I get _bridge variable?

**Comment by chevonc** [comment]
For those still hitting this issue, you can ensure the terminated app receives the native voice notif by calling `voipRegisteration` in your `AppDelegate.m` as earlier as possible, inside `didFnishLaunchingWithOptions`. Ex:

```
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  
  // register for voip notifs immediately to recv. notifs when app is killed - doing this from the JS side is too slow
  RNVoipPushNotificationManager* voipModule = [bridge moduleForClass:[RNVoipPushNotificationManager class]];
  [voipModule voipRegistration];
  
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"Guilded"
                                            initialProperties:nil];
```

This will call cause `didReceiveIncomingPushWithPayload` to properly be called on the native side when the app is terminated. 

*Be sure to remove any duplicated call to the JS equivalent of `voipRegistration` so two delegates aren't registered*

**Comment by Phineas** [comment]
@chevonc Tried adding this at same place you did to no avail, unfortunately.

**Comment by acro5piano** [comment]
I had the same issue but (maybe) thanks to @chevonc it was fixed.

**Comment by naderjavid** [comment]
thanks  every body!
i did every step but still doesn't work in background

[AppDelegate.txt](https://github.com/react-native-webrtc/react-native-voip-push-notification/files/5398062/AppDelegate.txt)

this is my AppDelegate file
please help.

**Comment by acro5piano** [comment]
@naderjavid I think you should try first a very simple example like this.

```
// --- Handle incoming pushes (for ios >= 11)
- (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(PKPushType)type withCompletionHandler:(void (^)(void))completion {
  NSString *uuid = @"44713e33-fa78-4ff5-8ec5-983e0832d1c6";
  NSString *callerName = @"Test";
  NSString *handle = @"handle";
  [RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES payload:nil];
}
```

**Comment by naderjavid** [comment]
> @naderjavid I think you should try first a very simple example like this. I got a call after I removed some extra code.
> 
> ```objc
> // --- Handle incoming pushes (for ios >= 11)
> - (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(PKPushType)type withCompletionHandler:(void (^)(void))completion {
>   NSString *uuid = @"44713e33-fa78-4ff5-8ec5-983e0832d1c6";
>   NSString *callerName = @"Test";
>   NSString *handle = @"handle";
>   [RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES payload:nil];
> }
> ```

thank you a lot my problem is solved

**Comment by naderjavid** [comment]
> @naderjavid I think you should try first a very simple example like this. I got a call after I removed some extra code.
> 
> ```objc
> // --- Handle incoming pushes (for ios >= 11)
> - (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(PKPushType)type withCompletionHandler:(void (^)(void))completion {
>   NSString *uuid = @"44713e33-fa78-4ff5-8ec5-983e0832d1c6";
>   NSString *callerName = @"Test";
>   NSString *handle = @"handle";
>   [RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES payload:nil];
> }
> ```

thank you @acro5piano my problem is solved!

**Comment by Andriiklymiuk** [comment]
have somebody have found viable solution?

**Comment by acro5piano** [comment]
@Andriiklymiuk in my case it doesn't work in the background state. Instead I decided to handle things in JS land(not in the native land). Now sure it works perfectly but it meets my need.

**Comment by zxcpoiu** [comment]
please test the fix in PR #69
